src/relay/pass/type_solver.cc

```diff
@@ -88,9 +88,56 @@ class TypeSolver::OccursChecker : public TypeVisitor {
   bool found_;
 };
 
+static Type GeneralizeShape(const std::vector<TensorType>& concrete_types) {
```

Comment:
Another nitpick but I think it could be relevant: I'm not sure a reference to "generalizing" is best here, since what it actually does is find the most _specific_ shape that fits all the type hints. Is there another name we can have for this logic?

Reply:
I think `generalization` is the right name for this function, it's the equivalent step as traditional type systems like [Hindley-Milner](http://okmij.org/ftp/ML/generalization.html#generalization). Instead of inserting polymorphic type vars in an expression, we replace those type vars with `Any`.

Reply:
I see your point @slyubomirsky honestly its like the lub of the type lattice, but I'm not sure if that is the best name either, will sleep on it.